### PR TITLE
[test.py] Add resource consumption metrics

### PR DESCRIPTION
**Pull request by xtrey**
This PR adds the possibility to gather resource consumption metrics. The collected metrics can be used to compare performance before and after specific changes aimed at increasing performance. Currently, this functionality works only in manual mode and this is just raw data. Later on, these metrics can be used in Jupyter notebook to analyze and visualize how the resources are used and can provide the insight on how to improve it. This [PR](https://github.com/scylladb/scylladb/pull/19832) is a first insight after gathering these metrics.

Add the possibility to gather resource consumption for the test.py execution. SQLite DB will be created with different performance metrics that will allow compare the resource consumption between changes.
The DB will be in the tmp directory that by default set to testlog. Across the runs, the DB will not be deleted, so each new run will just add information to the existing DB.
Parameter --get-metrics was added to switch on or off the metrics gathering. By default, it's switched on. 
The current limitation is that the cgroup statistics is done for tests itself without clusters because even that cluster is created by the test the architecture with cluster manager leaves them in the initial cgroup.

For example, to get the top 5 tests by memory peak, we can use the next query:
```SQL
SELECT t.test_name, t.mode, t.directory, m.memory_peak
FROM tests t
JOIN test_metrics m ON t.id = m.test_id
ORDER BY m.memory_peak DESC
LIMIT 5;
```
Here is the output example:
| test\_name | mode | directory | memory\_peak |
| :--- | :--- | :--- | :--- |
| test\_read\_repair | debug | topology\_custom | 119078912 |
| test\_change\_rpc\_address | debug | topology\_custom | 81608704 |
| test\_mv\_backlog | debug | topology\_custom | 76627968 |
| test\_replace | dev | topology\_custom | 59293696 |
| test\_mv\_backlog | dev | topology\_custom | 58101760 |

Closes: https://github.com/scylladb/qa-tasks/issues/1666

Closes: https://github.com/scylladb/qa-tasks/issues/1707

### CODE REVIEW

**Comment by mykaul** [comment]
Out of curiosity (as I did not follow / missed any discussion on the design of this): why not just set up a Prometheus node exporter and gather the data via it to a central Prometheus?

**Comment by xtrey** [comment]
> Out of curiosity (as I did not follow / missed any discussion on the design of this): why not just set up a Prometheus node exporter and gather the data via it to a central Prometheus?

We don't need it for every run, it will be mostly used for local runs. So using Prometheus, I think it's overkill. Additional use case of cgroups is the possibility to control the resources and simulation of lack of CPU, memory for the Scylla process under test.

**Comment by xtrey** [comment]
> Using Prometheus (which we have already for our on-prem hosts, btw) is an overkill, but setting a whole DB for it + collecting ourselves (in Python) the metrics in a non-standard way is better?

I'm using the SQLite db. It will be stored in testlog file as an artifact, no need to set up any new DB. Python has a built-in module to work with SQLite. AFAIK, there is not so easy to get the resources consumed in cgroup only.

**Comment by xtrey** [comment]
@mykaul here the [doc](https://docs.google.com/document/d/1iQr4qPo8JWH43mfIoxauerWMYc0tzbUamrRFH375Vbw/edit#heading=h.prspeijbjau9) with discussion

File: test/pylib/resource_gather.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+                _parse_cpu_stat(file, test_metrics)
+        return test_metrics
+
+    def write_metrics_to_db(self, metrics: TestMetricRecord, success: bool = False) -> None:
+        metrics.success = success
+        metrics_dict = asdict(metrics, filter=lambda attr, value: value is not None)
+        self.sqlite_writer.write_row(metrics_dict, METRICS_TABLE)
+
+
+def _parse_cpu_stat(file: TextIO, metrics: TestMetricRecord):
```
Comment (temichus) [inline test/pylib/resource_gather.py:93]: can be static method of the class

File: test.py [hunk 2]
```diff
@@ -1704,19 +1717,22 @@ async def main() -> int:
...
 
     signaled = asyncio.Event()
+    stop_event = asyncio.Event()
+    run_resource_watcher(options.gather_metrics, signaled, stop_event)
 
     setup_signal_handlers(asyncio.get_running_loop(), signaled)
 
     try:
         await run_all_tests(signaled, options)
+        stop_event.set()
```
Comment (temichus) [inline test.py:1735]: you need to wait for monitor_resources to finish as well as it has 2 seconds of sleep in the loop. also if it crashes you will be able to get stack trace only at the join

File: test/pylib/resource_gather.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+from attrs import define, field, asdict
+
+from test.pylib.db.model import TestMetricRecord, ResourceTimelineRecord
+from test.pylib.db.writer import DATE_TIME_TEMPLATE, SQLiteWriter, RESOURCE_TIMELINE_TABLE, DB_PATH, METRICS_TABLE
+
+CGROUP_INITIAL: Path = Path('/sys/fs/cgroup/initial')
+CGROUP_TESTS: Path = CGROUP_INITIAL.parent / 'tests'
+
+
+@define
```
Comment (temichus) [inline test/pylib/resource_gather.py:22]: can't see the advantage of using @define here. 

regular def __init__ will be more readable for me

File: test/pylib/resource_gather.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+    def put_process_to_cgroup(self, pid):
+        try:
+            with open(self.cgroup_path / 'cgroup.procs', "a") as cgroup:
+                cgroup.write(str(pid))
+        except ProcessLookupError:
+            logging.getLogger().log(
+                level=logging.WARNING,
+                msg=f"Process finished too quick. No metrics will be logged",
+            )
```
Comment (mlitvk) [inline test/pylib/resource_gather.py:66]: If the process doesn't exist, would this code trigger a `ProcessLookupError`?
I tried testing this and I got `OSError`

**Comment by avikivity** [comment]
> Add the possibility to gather resource consumption for the test.py execution. SQLite DB will be created with different performance metrics that will allow compare the resource consumption between changes.
> 
> Closes: [scylladb/qa-tasks#1666](https://github.com/scylladb/qa-tasks/issues/1666)
> 
> Closes: [scylladb/qa-tasks#1707](https://github.com/scylladb/qa-tasks/issues/1707)

Again, you need to explain why all of this is a good idea. Not just what we're doing.

File: test/pylib/db/__init__.py
Comment (avikivity) [inline test/pylib/db/__init__.py:1]: Why add an empty file?

File: test/pylib/db/model.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
+from attrs import define, AttrsInstance
```
Comment (avikivity) [inline test/pylib/db/model.py:1]: Missing: copyright/license blurb.

File: test/pylib/db/model.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+@define
+class TestMetricRecord(AttrsInstance):
+
+    architecture: str = None
+    mode: str = None
+    test_name: str = None
+    run_id: int = None
+    directory: str = None
+    user_usec: int = None
+    system_usec: int = None
```
Comment (avikivity) [inline test/pylib/db/model.py:13]: Suggest to use floats with units of seconds, it's more natural.

File: test/pylib/db/model.py [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    usage_usec: int = None
+    memory_peak: int = None
+    time_taken: int = None
```
Comment (avikivity) [inline test/pylib/db/model.py:16]: float

File: test/pylib/db/writer.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
+import sqlite3
```
Comment (avikivity) [inline test/pylib/db/writer.py:1]: copyright/license blurb.

Does this need an update to install-dependencies.sh?

File: test/pylib/resource_gather.py [hunk 9]
```diff
@@ -0,0 +1,166 @@
+import asyncio
```
Comment (avikivity) [inline test/pylib/resource_gather.py:1]: copyright/license

File: test/pylib/resource_gather.py [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+        if is_podman:
+            subprocess.run(['sudo', 'chown', '-R', f"{getpass.getuser()}:{getpass.getuser()}", '/sys/fs/cgroup'],
+                           check=True)
+        else:
+            subprocess.run(
+                [
+                    "sudo",
+                    "mount",
+                    "-o",
+                    "remount,rw,memory_recursiveprot,nsdelegate",
```
Comment (avikivity) [inline test/pylib/resource_gather.py:121]: Please add some comments about these options.

You're affecting the callers environment. Is this okay?

File: tools/toolchain/dbuild [hunk 11]
```diff
@@ -155,7 +155,9 @@ if [ -z "$is_podman" ]; then
        )
 else
     docker_common_args+=(
-        --userns keep-id)
+        --userns keep-id
+        --systemd=always
```
Comment (avikivity) [inline tools/toolchain/dbuild:159]: >               • Podman sets the default stop signal to SIGRTMIN+3.

Please verify that ctrl-C still works during build and test.py execution. If not we may need to override --stop-signal.

**Comment by avikivity** [comment]
> > Add the possibility to gather resource consumption for the test.py execution. SQLite DB will be created with different performance metrics that will allow compare the resource consumption between changes.
> > Closes: [scylladb/qa-tasks#1666](https://github.com/scylladb/qa-tasks/issues/1666)
> > Closes: [scylladb/qa-tasks#1707](https://github.com/scylladb/qa-tasks/issues/1707)
> 
> Again, you need to explain why all of this is a good idea. Not just what we're doing.

Also, add an example SELECT ... LIMIT with sorting by (say) CPU consumption as a demonstration of the benefit and of how to use the new feature.

**Review by temichus — COMMENTED** [review]
now test.py looks good. no if or for just additional method calls

ResourceGather is a real module that can be turned on/off.

spawned watcher properly finished

File: test/pylib/resource_gather.py (hunk 4 above)
Comment (xtrey) [reply test/pylib/resource_gather.py:66]: Double-checked and got `ProcessLookupError`. If you have `OSError`, please share the details here or in Slack. Maybe you are hitting some edge case and I need to handle it too.

File: test/pylib/db/writer.py (hunk 8 above)
Comment (xtrey) [reply test/pylib/db/writer.py:1]: There is no need to update it. The sqlite3 module is a part of the standard python library.

File: test/pylib/resource_gather.py (hunk 10 above)
Comment (xtrey) [reply test/pylib/resource_gather.py:121]: Removed these options as redundant in our case. 
This intended to be used in containers. Without toolchain this part will break.

File: test/pylib/resource_gather.py [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+            f.write(controllers)
+
+
+async def monitor_resources(cancel_event: asyncio.Event, stop_event: asyncio.Event, tmpdir: Path):
+    """Continuously monitors CPU and memory utilization."""
+    sqlite_writer = SQLiteWriter(tmpdir / DEFAULT_DB_NAME)
+    timeline_record = ResourceTimelineRecord()
+    while not cancel_event.is_set() and not stop_event.is_set():
+        timeline_record.cpu = psutil.cpu_percent(interval=1)
+        timeline_record.memory = psutil.virtual_memory().percent
```
[the remark this answers is not in the document]
Comment (xtrey) [reply test/pylib/resource_gather.py:150]: We are recording the consumption of memory in the system and CPU utilization in the system. For the cgroup we are recording the peak memory consumption and CPU usage in seconds. After discussion with @kostja @temichus we decided that this will be enough, at least as an initial step. We can already gather information better than we had, however, still not as granular as can be.
I have some concerns about monitoring memory consumption in cgroup. cgroup provides a controller to get current memory consumption, that means we need to spawn monitor process equal to threads we are executing. With boost tests, it's mostly not possible. Boost tests are executed by function and many of them last less than a second, so it will be difficult to get some meaningful data. Long-lasting tests should not be affected, but still, I think it's a matter of further thorough investigation.

[the remark this answers is not in the document]
Comment (avikivity) [reply test/pylib/resource_gather.py:150]: > We are recording the consumption of memory in the system and CPU utilization in the system. For the cgroup we are recording the peak memory consumption and CPU usage in seconds. After discussion with @kostja @temichus we decided that this will be enough, at least as an initial step. 

Maybe it's enough, but I don't understand what you're recording.

> We can already gather information better than we had, however, still not as granular as can be. I have some concerns about monitoring memory consumption in cgroup. cgroup provides a controller to get current memory consumption, that means we need to spawn monitor process equal to threads we are executing.

I'm sure you can read the cgroup report from outside the cgroup.

>  With boost tests, it's mostly not possible. Boost tests are executed by function and many of them last less than a second, so it will be difficult to get some meaningful data. Long-lasting tests should not be affected, but still, I think it's a matter of further thorough investigation.

There's getrusage().ru_maxrss. This works for a terminated process.

File: test/pylib/db/writer.py (hunk 8 above)
Comment (avikivity) [reply test/pylib/db/writer.py:1]: Nice

File: test/pylib/db/__init__.py
Comment (avikivity) [reply test/pylib/db/__init__.py:1]: Ok. Please add license/copyright blurbs.

File: test/pylib/resource_gather.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (avikivity) [reply test/pylib/resource_gather.py:150]: > > Maybe it's enough, but I don't understand what you're recording.
> 
> I'll update the docstring to the method

I mean, I understand that it's a percentage of virtual memory, just not how it's useful.

> 
> > I'm sure you can read the cgroup report from outside the cgroup.
> 
> It's not a problem to read it outside the cgroup. The issue there is that metrics exist only in real time. Once the process is done, memory will show 0 usage.


> 
> > There's getrusage().ru_maxrss. This works for a terminated process.
> 
> Please elaborate. I don't see how I can provide PID to the method to get the data and test.py is executed only once, so ru_maxrss will show maximum not only for the latest test, but for all tests that were already executed.

test.py execs ./build/mode/test/boost/whatever  and waits for it. The wait4() system call (and presumable the Python ecosystem) returns a struct rusage. The struct rusage contains ru.maxrss that says what the peak memory consumption was.

This doesn't work for forked children, but our boost tests don't fork.

It's also possible to assume some value like 50MB. If a test completes so quickly that we don't get to sample it even once, we can assume it didn't allocate a lot of memory.

File: tools/toolchain/dbuild (hunk 11 above)
Comment (xtrey) [reply tools/toolchain/dbuild:159]: It's not working on master branch either. Tried to add `--stop-signal=SIGTERM` and `--stop-signal=SIGKILL`, but still no luck.

File: test/pylib/db/__init__.py
Comment (xtrey) [reply test/pylib/db/__init__.py:1]: Added in the files where we have code. Do you want me to add it to this empty file as well?

File: tools/toolchain/dbuild (hunk 11 above)
Comment (avikivity) [reply tools/toolchain/dbuild:159]: It's working for me with podman, but only interactive mode (invoking dbuild without parameters).

Comment (xtrey) [reply tools/toolchain/dbuild:159]: It works when I'm already in interactive mode, when I start the build or test.py and then cancel them.

**Review by temichus — APPROVED** [review]

File: test/pylib/resource_gather.py [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+        super().__init__(test, tmp_dir)
+        self.sqlite_writer = SQLiteWriter(self.db_path)
+
+    def make_cgroup(self):
+        os.makedirs(self.cgroup_path, exist_ok=True)
+
+    def put_process_to_cgroup(self, pid):
+        try:
+            with open(self.cgroup_path / 'cgroup.procs', "a") as cgroup:
+                cgroup.write(str(pid))
```
[the remark this answers is not in the document]
Comment (xtrey) [reply test/pylib/resource_gather.py]: Rewrote to use preexec_fn

File: test/pylib/resource_gather.py (hunk 9 above)
Comment (xtrey) [reply test/pylib/resource_gather.py:7]: done

File: test/pylib/db/model.py (hunk 6 above)
Comment (xtrey) [reply test/pylib/db/model.py]: done

File: test/pylib/db/model.py (hunk 5 above)
Comment (xtrey) [reply test/pylib/db/model.py]: done

**Comment by avikivity** [comment]
@xtrey is this ready for re-review? What are the changes?

**Comment by xtrey** [comment]
> @xtrey is this ready for re-review? What are the changes?

Yes, it's ready.

Resolved all your comments: with int->float, usec -> sec, create pid before starting the process, reorganized the tables in DB, added monitoring for memory consumption in cgroup.

**Comment by avikivity** [comment]
> > @xtrey is this ready for re-review? What are the changes?
> 
> Yes, it's ready.
> 
> Resolved all your comments: with int->float, usec -> sec, create pid before starting the process, reorganized the tables in DB, added monitoring for memory consumption in cgroup.

Usually, one notes the list of changes, otherwise the reviewer has to redo the review from scratch (given the scope of the changes I probably have to do that anyway).

File: test.py [hunk 14]
```diff
@@ -36,6 +36,7 @@
 from test.pylib.artifact_registry import ArtifactRegistry
 from test.pylib.host_registry import HostRegistry
```
Comment (avikivity) [inline test.py:37]: > [test.py] Add resource consumption metrics

Please enrich the patch changelog with details.

File: test.py [hunk 15]
```diff
@@ -1263,7 +1280,9 @@ def report_error(error):
                 print(test.shortname, end=" ")
                 report_error("Test was cancelled: the parent process is exiting")
         except Exception as e:
-            report_error("Failed to run the test:\n{e}".format(e=e))
+            with open(f'{options.tmpdir}/stack.out', 'w+') as file:
+                traceback.print_exc(file=file)
```
Comment (avikivity) [inline test.py:1284]: I don't understand the purpose of this change.

File: test/pylib/db/model.py [hunk 16]
```diff
@@ -0,0 +1,42 @@
+#
+# Copyright (C) 2022-present ScyllaDB
```
Comment (avikivity) [inline test/pylib/db/model.py:2]: It's 2024.

File: test/pylib/db/writer.py [hunk 17]
```diff
@@ -0,0 +1,154 @@
+#
+# Copyright (C) 2022-present ScyllaDB
```
Comment (avikivity) [inline test/pylib/db/writer.py:2]: 2024

File: test/pylib/db/writer.py [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+        """
+        Initializes the SQLWriter object.
+
+        Args:
+            database_path: Path to the SQLite database file.
+        """
+        self.lock = asyncio.Lock()
+        self.conn = sqlite3.connect(database_path)
+        self.cursor = self.conn.cursor()
+        self.cursor.execute('PRAGMA foreign_keys=ON')
```
Comment (avikivity) [inline test/pylib/db/writer.py:91]: let's also issue 'PRAGMA sychronous=off' to disable fsync; fsync can stall other coroutines.

File: test/pylib/resource_gather.py [hunk 19]
```diff
@@ -0,0 +1,195 @@
...
+                test_name=test.shortname
+            ),
+            TESTS_TABLE)
+
+    def make_cgroup(self):
+        os.makedirs(self.cgroup_path, exist_ok=True)
+
+    def get_test_metrics(self) -> Metric:
+        test_metrics: Metric = Metric(test_id=self.test_id)
+        test_metrics.time_taken = round(int(self.test.time_end - self.test.time_start))
```
Comment (avikivity) [inline test/pylib/resource_gather.py:78]: Why round?

File: test/pylib/resource_gather.py [hunk 20]
```diff
@@ -0,0 +1,195 @@
...
+                timeline_record = CgroupMetric(
+                    test_id=self.test_id,
+                    memory=int(memory_current.read()),
+                    timestamp=datetime.now()
+                )
+
+                self.sqlite_writer.write_row(timeline_record, CGROUP_MEMORY_METRICS_TABLE)
+
+            # Control the frequency of updates, for example, every 2 seconds
+            await asyncio.sleep(1)
```
Comment (avikivity) [inline test/pylib/resource_gather.py:117]: Since you have the memory.peak file, any reason to have this coroutine?

File: test/pylib/resource_gather.py [hunk 21]
```diff
@@ -0,0 +1,195 @@
...
+            subprocess.run(
+                [
+                    "sudo",
+                    "mount",
+                    "-o",
+                    "remount,rw",
+                    "/sys/fs/cgroup",
+                ],
+                check=True,
+            )
```
Comment (avikivity) [inline test/pylib/resource_gather.py:153]: It's unfortunate to run so many processes for every test case, but we can tolerate it for now.

File: test/pylib/resource_gather.py [hunk 22]
```diff
@@ -0,0 +1,195 @@
...
+        timeline_record = SystemResourceMetric(
+            cpu=psutil.cpu_percent(interval=0.1),
+            memory=psutil.virtual_memory().percent,
+            timestamp=datetime.now()
+        )
+
+        sqlite_writer.write_row(timeline_record, SYSTEM_RESOURCE_METRICS_TABLE)
+
+        # Control the frequency of updates, for example, every 2 seconds
+        await asyncio.sleep(2)
```
Comment (avikivity) [inline test/pylib/resource_gather.py:185]: This is the second such coroutine, why two?

**Comment by avikivity** [comment]
Looks reasonable. Please add an example of using sqlite3 to query the top 5 resource users and the output to the patch changelog.

File: test.py (hunk 15 above)
Comment (xtrey) [reply test.py:1284]: This is leftovers after debugging. I'll revert this.

File: test.py [hunk 23]
```diff
@@ -1305,6 +1324,7 @@ def parse_cmd_line() -> argparse.Namespace:
         help="""Path to temporary test data and log files. The data is
         further segregated per build mode. Default: ./testlog.""",
     )
+    parser.add_argument("--gather_metrics", action=argparse.BooleanOptionalAction, default=False)
```
[the remark this answers is not in the document]
Comment (xtrey) [reply test.py:1327]: The main reason is to not break the flow if somebody tries to execute the test.py outside the toolchain. Additionally, I doubt that is beneficial in the current state to switch on since it just metric gathering, and they are useless if nobody analyzing them.

[the remark this answers is not in the document]
Comment (avikivity) [reply test.py:1327]: Why would it break the flow?

I like to keep CI using defaults, so developers aren't surprised when something works on their machine but doesn't work on CI.

We can add ~/.config/scylladb/scylladb-test-py.conf so people can adjust their configuration to taste.

File: test.py [hunk 24]
```diff
@@ -1776,19 +1796,24 @@ async def main() -> int:
...
     signaled = asyncio.Event()
+    stop_event = asyncio.Event()
+    resource_watcher = run_resource_watcher(options.gather_metrics, signaled, stop_event, options.tmpdir)
 
     setup_signal_handlers(asyncio.get_running_loop(), signaled)
 
     try:
         await run_all_tests(signaled, options)
+        stop_event.set()
+        async with asyncio.timeout(5):
```
[the remark this answers is not in the document]
Comment (avikivity) [reply test.py:1815]: Thanks. In Seastar we can't cancel things, so we don't have this pattern.

[the remark this answers is not in the document]
Comment (avikivity) [reply test.py:1815]: But, since this is just a safety mechanism, I suggest to increase the timeout even more (since we're trying to load the machine), and report when it triggered.

File: test.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (xtrey) [reply test.py:1327]: The idea was to switch it on upon need. Implemented logic will not work outside the toolchain. So with default set to true and using just test.py will not work because we have no power to manipulate the cgroups in the root. Even if I rewrite the logic to start from the cgroup where the process is, there still can be an issue that this cgroup will be owned by root, and we will not be able to create another cgroup or put the PID to the desired one.

File: test.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (xtrey) [reply test.py:1815]: It's not cancelling, it's stopping of the monitoring. We are using this event as a flag that we need to stop the loop. I think this timeout is enough since it double the time of the sleep in waiting process. The only thing that I should implement is loggin an error in case of timeout happens, rather than failing the whole process. In the worst case, we will lose only the last metric with a CPU and RAM utilization, the rest should be already written.

File: test/pylib/resource_gather.py (hunk 19 above)
Comment (xtrey) [reply test/pylib/resource_gather.py:78]: Missed to change it. It's redundant, so I'll remove it.

File: test/pylib/resource_gather.py (hunk 20 above)
Comment (xtrey) [reply test/pylib/resource_gather.py:117]: For better understanding of the memory use. For example, we have one test in a file that consumes a lot of memory, but the rest of the test do not need so much memory. With only peak memory, we will lose this information, and we will not be able to schedule the test correctly.

File: test/pylib/resource_gather.py [hunk 25]
```diff
@@ -0,0 +1,195 @@
...
+    def _parse_cpu_stat(file: TextIO, metrics: Metric) -> None:
+        # Map the values from cpu.state to the model. Keys are values from cpu.stat
+        stats = {'user_usec': 'user_sec', 'system_usec': 'system_sec', 'usage_usec': 'usage_sec'}
+        for line in file.readlines():
+            stat, value = line.split(' ')
+            if stat in stats.keys():
+                setattr(metrics, stats[stat], float(value) / 1_000_000)
+
+
+def get_resource_gather(is_switched_on, test, tmpdir: str) -> ResourceGather:
```
[the remark this answers is not in the document]
Comment (xtrey) [reply test/pylib/resource_gather.py:129]: I'll add an annotation to the is_switched_on, but the test will be without it. Class Test is in test.py, and it's outside python module and test.py importing the resource_gather. If I'll try to add Test in resource_gather it will lead to a recursive import, and python will raise an exception.

File: test/pylib/resource_gather.py (hunk 22 above)
Comment (xtrey) [reply test/pylib/resource_gather.py:185]: One monitoring coroutine to gather system CPU and RAM utilization, the second one is to monitor the cgroup memory usage that is spawned for each test.

File: test/pylib/resource_gather.py (hunk 21 above)
Comment (xtrey) [reply test/pylib/resource_gather.py:153]: It's an architecture of the test.py. Each directory is counted as a suite, and it will have its own python process spawned. Worse situation with boost test, where each test function will have its own spawned python process. It was done so to decrease the time of the execution of the boost tests. One of the boost tests takes too much time if we execute it by file.

File: test.py (hunk 14 above)
Comment (xtrey) [reply test.py:37]: By changelog, you mean to add information to the PR description? Because I was unable to find the changelog file dedicated to test.py.

File: test.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (avikivity) [reply test.py]: > The idea was to switch it on upon need. Implemented logic will not work outside the toolchain. 

Why not? Everything else does. Please don't write code so fragile it only works in one scenario.

> So with default set to true and using just test.py will not work because we have no power to manipulate the cgroups in the root. Even if I rewrite the logic to start from the cgroup where the process is, there still can be an issue that this cgroup will be owned by root, and we will not be able to create another cgroup or put the PID to the desired one.

I don't understand why. Podman is able to create cgroups without being root. So should we.

File: test/pylib/resource_gather.py (hunk 20 above)
Comment (avikivity) [reply test/pylib/resource_gather.py:117]: I don't think we'll be able to utilize this information. The execution history depends on what other tests ran in parallel. The next execution will have different timing.

File: test/pylib/resource_gather.py (hunk 21 above)
Comment (avikivity) [reply test/pylib/resource_gather.py:153]: For boost tests, once we gather history, we can run all test cases that take a short amount of time in a single execution.

File: test/pylib/resource_gather.py (hunk 22 above)
Comment (avikivity) [reply test/pylib/resource_gather.py:185]: I see. Do we need to keep system resource utilization in sqlite? It can't help future schedules.

File: test/pylib/resource_gather.py (hunk 20 above)
Comment (xtrey) [reply test/pylib/resource_gather.py:117]: Our tests executed in alphabetical order. So every run has the same order of execution. Until we start shuffling them, we can compare this data between runs.

File: test/pylib/resource_gather.py (hunk 22 above)
Comment (xtrey) [reply test/pylib/resource_gather.py:185]: It can help to analyze the whole picture of where we're still not utilizing the RAM or CPU and where we can add additional thread.

File: test.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (xtrey) [reply test.py]: > Why not? Everything else does. Please don't write code so fragile it only works in one scenario.

It's not fragile. It works in the podman and in the docker. Generally speaking, in the environment where we have control. Outside the container there can be any configuration, and it will be hardly possible to cover all of them.

> I don't understand why. Podman is able to create cgroups without being root. So should we.

We can create the cgroups as well without root in the cgroups where the user is an owner. We can assume that the process where tetst.py is started located in the cgroup with correct permission, but this is not always true. In that case, we need to find the cgroup with correct permissions. After that, the flow will be actually the same. But finding the correct cgroup can be tricky a little bit.

**Comment by xtrey** [comment]
Added handling the ```asyncio.exceptions.CancelledError``` to not fail the whole job.

File: test.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (avikivity) [reply test.py]: Does this not work?

```console
$ cat /proc/$$/cgroup 
0::/user.slice/user-1000.slice/user@1000.service/app.slice/app-org.gnome.Terminal.slice/vte-spawn-aff2c052-ef9b-4023-afef-3bd36a81fa19.scope
```

[the remark this answers is not in the document]
Comment (xtrey) [reply test.py]: It works, but for example, my foot terminal instance will be in the cgroup where the root is owner. So in most cases it will be easy to go get cgroup through the /proc/self/cgroup, but still, we cannot rely fully on it.

**Comment by avikivity** [comment]
I get this on my Fedora 40 workstation:

```
$ ./test.py --mode dev
Traceback (most recent call last):
  File "/home/avi/scylla-maint/./test.py", line 2097, in workaround_python26789
    code = await main()
           ^^^^^^^^^^^^
  File "/home/avi/scylla-maint/./test.py", line 1791, in main
    setup_cgroup(options.gather_metrics)
  File "/home/avi/scylla-maint/test/pylib/resource_gather.py", line 161, in setup_cgroup
    directory.mkdir()
  File "/usr/lib64/python3.12/pathlib.py", line 1311, in mkdir
    os.mkdir(self, mode)
PermissionError: [Errno 13] Permission denied: '/sys/fs/cgroup/initial'
```

**Comment by xtrey** [comment]
> I get this on my Fedora 40 workstation:

Added possibility to run without toolchain. However, there still be some caveats and assumptions to make it work. The terminal where we are executing, starting the process should be located in cgroup owned by the current user.

File: test/pylib/resource_gather.py [hunk 26]
```diff
@@ -0,0 +1,219 @@
...
+        return ResourceGatherOn(test, tmpdir)
+    else:
+        return ResourceGatherOff(test, tmpdir)
+
+
+def setup_cgroup(is_required: bool) -> None:
+    if is_required:
+        # check where the process is executed in podman or in docker
+        is_podman = os.access("/run/.containerenv", os.F_OK)
+        is_docker = os.getpid() == 1
```
Comment (avikivity) [inline test/pylib/resource_gather.py:156]: I don't think this is correct. If you launch the container in interactive mode, bash will have pid 1, and test.py will have another pid. Similar if you launch it with something like ./tools/toolchain/dbuild bash -c 'ninja dev-build && ./test.py --mode dev` as I sometimes do.

File: test/pylib/resource_gather.py [hunk 27]
```diff
@@ -0,0 +1,219 @@
...
+        elif is_docker:
+            subprocess.run(
+                [
+                    "sudo",
+                    "mount",
+                    "-o",
+                    "remount,rw",
+                    "/sys/fs/cgroup",
+                ],
+                check=True,
```
Comment (avikivity) [inline test/pylib/resource_gather.py:169]: Do the two container environments really need such different methods to give the current user permissions?

Perhaps we should revert 34e3302c017? Though the equivalent was necessary, at least with older docker.

Perhaps recent docker can remap files written as root to the original user.

File: test/pylib/resource_gather.py (hunk 26 above)
Comment (xtrey) [reply test/pylib/resource_gather.py:156]: It's fragile. But I found a better way, I was not aware that we switched to use by default podman instead of docker.

Comment (xtrey) [reply test/pylib/resource_gather.py:156]: The behavior of cgroups is different for podman, docker and local. In docker cgroups mounted with RO permission, so we need to remount it. In podman we need to own cgroup to be able to use them how we want. Locally we cannot be user and should rely on the cgroup of the current process and hoping that user has permissions to write to it.
`container=podman` injected by systemd, so docker do not have it, and if we want to have something like it, we have to pass it manually. The best way so far that I was able to find is to check files `/run/.containerenv` for podman and `/.dockerenv`

File: test/pylib/resource_gather.py (hunk 27 above)
Comment (xtrey) [reply test/pylib/resource_gather.py:169]: Unfortunately, yes. It is still necessary to remount the cgroup for Docker. Change that you mentioned affects podman, not docker.

**Comment by xtrey** [comment]
Removed `--systemd=always` for podmand since we migrated to a newer version of Fedora and podman, so no need for this workaround for the bug https://github.com/containers/podman/issues/20073
Changed the way of distinguishing that the process in the docker container

**Comment by xtrey** [comment]
Added try-catch when stopping the cgroup monitoring.
This is flaky, and I can't reproduce it locally to check that the fix is working. Even with repeat 100 there were no log messages regarding the `CancelledError` or `TimeoutError`
image

**Comment by xtrey** [comment]
Added cleaning the cgroup after the test

File: test/pylib/resource_gather.py (hunk 26 above)
Comment (avikivity) [reply test/pylib/resource_gather.py]: What I meant is: check if it's mounted RO, if so remount it RW. This way, if docker/podman changes its behavior, it doesn't break dbuild.

File: test/pylib/resource_gather.py (hunk 27 above)
Comment (avikivity) [reply test/pylib/resource_gather.py:175]: So here, look at /proc/mounts, check if the /sys/fs/cgroup entry is `ro`, and change it to rw if it is. This way, changes in the docker/podman runtime don't affect us.

Comment (xtrey) [reply test/pylib/resource_gather.py:175]: I've modified the code to check that we in a docker container and remount only for docker. So current code does not affect local run or execution in podman.

File: test/pylib/resource_gather.py (hunk 26 above)
Comment (avikivity) [reply test/pylib/resource_gather.py]: podman/docker aren't monolithic tools. They call into runtimes, which can be different based on the OS, and the runtimes change independently of podman/docker (crun/runc/whatever).

**Comment by xtrey** [comment]
Remount cgroup2 will be done for docker only and when it only mounted in RO mode.

File: test/pylib/resource_gather.py (hunk 27 above)
Comment (avikivity) [reply test/pylib/resource_gather.py:186]: Ok. It would be better to only check for is_cgroup_ro, not is_docker. The idea is that is docker or podman change their behavior in the future, and we build an old release, we don't have to fight with this script.

Comment (xtrey) [reply test/pylib/resource_gather.py:186]: podman not allow remounting. If podman changes something here, we will need to modify our approach. This was my main point, we cannot predict changes in docker/podman so we cannot write a code before these changes will arrive. For docker remounting seems redundant, with privileged mode it will mount RW, without it will be RO, but there will be no way to remount it with sudo.
